### PR TITLE
[docs] improve gitpod.io ddev-gitpod-launcher instructions with a bookmarklet

### PR DESCRIPTION
**Pull request by tyler36**
## The Problem/Issue/Bug:
Gitpod.io Usage list 3 ways of accessing DDEV via Gitpod.

1. Requires manually installing brew
2. Required typing a URL (error prone)
3. Required Chrome (or a browser that supports Chrome extentions)

## How this PR Solves The Problem:
This PR adds a bookmarklet option to open ddev-gitpod. 

This has 2 main benefits:
- grants the benefits of "DDEV Gitpod Launcher Chrome Extension", but for (almost) all browsers including Safari
- native browser keyboard shortcuts can be used

I say _almost _ all mostly because I'm not too familiar with browsers outside the top 4-5 popular ones. 

I am unable to test Safari, but a quick search says it supports bookmarklets.
I did test on Chrome, Firefox & Opera.

## What it does
Basically, if you are a github page, it re-writes the URL to a ddev-gitpod, else it redirects you back to the same site (incase you accidently clicked it)

## Automated Testing Overview:
[skip ci][ci skip]

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3451"></a>

### CODE REVIEW

**Comment by tyler36** [comment]
I thought I remembered assigning keyboard short-cuts to specific bookmarks in the distant past. That would make it super easy to hit `ctrl+q` to activate but I can't seem to find the feature anymore.

**Review by rfay — CHANGES_REQUESTED** [review]
Thanks for this!

It's not quite working out yet,
ddev_gitpod_md_at_0f689022c0e06d52cf6f0fa439fd29c354b8bb06_·_drud_ddev
 
Although I'm pretty sure what you're trying to do is going to be great, I'm not sure I understand it. You're just providing a sample link as a bookmark so people can edit with their own repo?

**Comment by tyler36** [comment]
It seems the code is being stripped out of the markdown. I tried to add it via an edit method but it didn't work that way either.

## Tests

Markdown:
[Github -> ddev-gitpod](javascript:window.location.href=window.location.href.includes('https://github.com/')?%27https://gitpod.io/#DDEV_REPO='+encodeURIComponent(window.location.href)+',DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/':window.location.href)

```markdown
[Github -> ddev-gitpod](javascript:window.location.href=window.location.href.includes('https://github.com/')?%27https://gitpod.io/#DDEV_REPO='+encodeURIComponent(window.location.href)+',DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/':window.location.href)
```

In "normal" html you can use an anchor. Markdown parses are _suppose_ to ignore HTML but it seems the Github parser & readthedocs parser is removing them.

 Anchor: <a href="javascript:window.location.href=window.location.href.includes('https://github.com/')?%27https://gitpod.io/#DDEV_REPO='+encodeURIComponent(window.location.href)+',DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/':window.location.href">Github -> ddev-gitpod</a>

```html
<a href="javascript:window.location.href=window.location.href.includes('https://github.com/')?%27https://gitpod.io/#DDEV_REPO='+encodeURIComponent(window.location.href)+',DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/':window.location.href">Github -> ddev-gitpod</a>
```


## Alternatives

- Include the raw javascript. Users can make their own bookmark from it. It's not pretty though.

```js
javascript: window.location.href = window.location.href.includes('https://github.com/') ? %27https://gitpod.io/#DDEV_REPO=' + encodeURIComponent(window.location.href) + ',DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/' : window.location.href
```

- If the [ddev-gitpod-launcher](https://drud.github.io/ddev-gitpod-launcher/) website is HTML based, it could be included there instead.



## Explanation

Bookmarks are usually URLs. However, they can be also be javascript. 

In this PR, it should be redirecting the current Github page to a ddev-gitpod page, via javascript. I use to use something similar to open an online VScode IDE before Microsoft included it by default (hit `.` on a Github repo page).

While the extension is great, both Chrome & Firefox have a history of "popular" extensions being bought and subverted.  Extensions by nature are browser specific. If I can use a bookmark instead, it's my prefered go-to.

Checkout out [41 Bookmarklets To Boost Your Productivity In Chrome, Firefox & Safari](https://www.thetechbasket.com/most-useful-bookmarklets/) for some other examples.

**Comment by rfay** [comment]
I guess there are a probably some hurdles.

1.  The first one, that I mistakenly didn't make it over, is that github is never going to render anything with js in it. But we may not care.
2. The second hurdle is doing the right thing with mkdocs to get it to work right. You can test that by just installing mkdocs and `mkdocs serve` in the root of the repo. Right now, the bookmarklet that does get generated is 
> javascript: window.location.href %3D window.location.href.includes(%27https://github.com/%27) %3F %27https://gitpod.io/%23DDEV_REPO%3D%27 %2B encodeURIComponent(window.location.href) %2B %27,DDEV_ARTIFACTS%3D/https://github.com/drud/ddev-gitpod-launcher/%27 : window.location.href
3. The third hurdle is making sure it works on readthedocs. I'll bet once it works with mkdocs it will be fine with readthedocs.

**Comment by tyler36** [comment]
OK ... I updated the link.

## Test
- I ran a python container with mkdocs and the required plugins. 
```shell
pip install mkdocs mkdocs-material mdx_truly_sane_lists mkdocs-git-revision-date-localized-plugin
```

- started a mkdocs server

```shell
mkdocs serve
```

- navigated to the [gipod page](http://127.0.0.1:8000/users/topics/gitpod/)
- dragged the link to Chrome bookmark bar
- navigrated Chrome to a [github repo](https://github.com/drud/ddev)
- click bookmarklet
- works. I'm redirected the ddev-gitpod page

**Comment by rfay** [comment]
I tried this and failed. I visited https://github.com/rfay/randyfay.com and clicked the link it generated:

`https://gitpod.io/#DDEV_REPO=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com,DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/`

ddev-gitpod-launcher would have generated:
`https://gitpod.io/#DDEV_REPO=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com,DDEV_ARTIFACTS=https%3A%2F%2Fgithub.com%2Fdrud%2Fd9simple-artifacts/https://github.com/drud/ddev-gitpod-launcher/
`

I assume that the `DDEV_ARTIFACTS=/` is the problem? It can be a nonexistent URL, but it should at least be a URL.

Starting_—_Gitpod

**Comment by rfay** [comment]
BTW, I think your approach here is probably way better than the chrome extension and will invalidate the whole idea of the chrome extension. Much appreciated!

**Comment by tyler36** [comment]
Hmmm ... just testing it with https://github.com/rfay/randyfay.com

Generated link:
https://gitpod.io/#DDEV_REPO=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com,DDEV_ARTIFACTS=/https://github.com/drud/ddev-gitpod-launcher/

It appears to work but the browser shows an error, which I assume is because of the missing `DDEV_ARTIFACTS`:
"
PDOException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'db.semaphore' doesn't exist: SELECT expire, value FROM {semaphore} WHERE name = :name; Array ( [:name] => variable_init ) in lock_may_be_available() (line 167 of /var/www/html/docroot/includes/lock.inc).
"

**Comment by tyler36** [comment]
https://drud.github.io/ddev-gitpod-launcher/

- type "https://github.com/rfay/randyfay.com" into git repo
- hit tab to move to next field
- "git repository with artifacts" is updated to "https://github.com/rfay/randyfay.com-artifacts"
- Button URL and "Link used" shows: `https://gitpod.io/#DDEV_REPO=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com,DDEV_ARTIFACTS=https%3A%2F%2Fgithub.com%2Fdrud%2Fd9simple-artifacts/https://github.com/drud/ddev-gitpod-launcher/`

Is that correct?
It appears to include `https%3A%2F%2Fgithub.com%2Fdrud%2Fd9simple-artifacts`. Shouldn't the artifacts repo include `https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com-artifacts`?

**Comment by tyler36** [comment]
OK, the link should now default the repo + '-artifacts` for DDEV artifacts.

- Visiting https://github.com/rfay/randyfay.com
- Generates `https://gitpod.io/#DDEV_REPO=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com,DDEV_ARTIFACTS=https%3A%2F%2Fgithub.com%2Frfay%2Frandyfay.com-artifacts/https://github.com/drud/ddev-gitpod-launcher/`

Gitpod boots althought the simple browser shows "Base table or view not found: 1146 Table 'db.semaphore' doesn't exist"

**Comment by rfay** [comment]
Yeah, there isn't any artifacts repo for randyfay.com repo. So it just fails to load it on the way up, which is fine.

**Review by rfay — APPROVED** [review]
Absolutely awesome! Wish I'd talked with you before spending all that time on the ddev-gitpod-launcher browser plugin!

**Comment by tyler36** [comment]
I know Chrome extension _definately_ allow keyboard shortcuts in the `manifest.json` which is nice-to-have feature.  

```
"commands": {
	"firstCommand": {
    	"suggested_key": "CTRL+F",
        "description": "My first command"
    }
}
```

From a maintainability & browser compability standpoint, I'm glad we can use a bookmarklet.

**Comment by tyler36** [comment]
I saw you changed the wording:

```diff
- When you are on a Github repository
+ When you are on a git repository
```

The bookmarklet, checks if you are on `github.com`. To me, this change implies it works on `https://bitbucket.org/` or `https://gitlab.com/` which is _not_ the case.

**Comment by rfay** [comment]
Oh, shucks. Thought it was generic. That's OK.  Should we struggle to make it more generous, or just change the wording back?
